projects/Magnetosphere/Magnetosphere.cpp

```diff
@@ -593,183 +594,288 @@ namespace projects {
-			  mpiGrid.refine_completely(myCell);			  
-		       }
-
-		    // Check if cell is within the tail box
-		    if ((xyz[0]>refine_L3tailxmin) && (xyz[0]<refine_L3tailxmax) &&
-			(abs(xyz[1])<refine_L3tailwidth) && (abs(xyz[2])<refine_L3tailheight))
-		       {
-			  CellID myCell = mpiGrid.get_existing_cell(xyz);
-			  // Check if the cell is tagged as do not compute
-			  mpiGrid.refine_completely(myCell);
-		       }
-
- 		 }
-	      }
-	   }
-	   refinedCells = mpiGrid.stop_refining(true);
-	   if(myRank == MASTER_RANK) std::cout << "Finished third level of refinement" << endl;
-#ifndef NDEBUG
-	   if(refinedCells.size() > 0) {
-	      std::cout << "Rank " << myRank << " refined " << refinedCells.size() << " cells. " << std::endl;
-	   }
-#endif
-	   
-	   mpiGrid.balance_load();
-     }
-
-     if (P::amrMaxSpatialRefLevel > 3) {
-	// L4 refinement.
-	   for (uint i = bw4; i < 8*P::xcells_ini-bw4; ++i) {
-	      for (uint j = bw4; j < 8*P::ycells_ini-bw4; ++j) {
-		 for (uint k = bw4; k < 8*P::zcells_ini-bw4; ++k) {
-		    
-		    std::array<double,3> xyz;
-		    xyz[0] = P::xmin + (i+0.5)*0.125*P::dx_ini;
-		    xyz[1] = P::ymin + (j+0.5)*0.125*P::dy_ini;
-		    xyz[2] = P::zmin + (k+0.5)*0.125*P::dz_ini;
-                    
- 		    Real radius2 = (xyz[0]*xyz[0]+xyz[1]*xyz[1]+xyz[2]*xyz[2]);
-
-		    // Check if cell is within the nose cap
-		    if ((xyz[0]>refine_L4nosexmin) && (radius2<refine_L4radius*refine_L4radius))
-		       {
-			  CellID myCell = mpiGrid.get_existing_cell(xyz);
-			  // Check if the cell is tagged as do not compute
-			  mpiGrid.refine_completely(myCell);			  
-		       }
-
- 		 }
-	      }
-	   }
-	   refinedCells = mpiGrid.stop_refining(true);
-	   if(myRank == MASTER_RANK) std::cout << "Finished fourth level of refinement" << endl;
-#ifndef NDEBUG
-	   if(refinedCells.size() > 0) {
-	      std::cout << "Rank " << myRank << " refined " << refinedCells.size() << " cells. " << std::endl;
-	   }
-#endif
-	   
-	   mpiGrid.balance_load();
-     }
-
-     return true;
+      #pragma omp parallel for
+      for (int j = 0; j < cells.size(); ++j) {
+         CellID id = cells[j];
+         std::array<double,3> xyz = mpiGrid.get_center(id);
+         SpatialCell* cell = mpiGrid[id];
+         int refLevel = mpiGrid.get_refinement_level(id);
+         Real r2 = pow(xyz[0], 2) + pow(xyz[1], 2) + pow(xyz[2], 2);
+
+         bool refine = false;
+         if (r2 < ibr2) {
+            // Skip refining, we shouldn't touch borders when reading restart
+            continue;
+         } else if (cell->parameters[CellParams::ALPHA] > refineTreshold) {
+            if (canRefine(xyz, refLevel)) {
+               #pragma omp critical
+               mpiGrid.refine_completely(id);
+            }
+         }
+      }
+
+      cells = mpiGrid.stop_refining();
+
+      #pragma omp parallel for
+      for (int j = 0; j < cells.size(); ++j) {
+         CellID id = cells[j];
+         *mpiGrid[id] = *mpiGrid[mpiGrid.get_parent(id)];
```

Comment:
Is this really thread-safe, if used in a parallel region like this?

Reply:
Parallelization removed.